jscomp/core/js_pass_flatten.ml

```diff
@@ -69,36 +68,36 @@ let flatten_map =
           end
       | Return {expression_desc = Cond (a,b,c);  comment}
         -> 
-          { statement_desc = If (a, [self#statement (S.return_stmt b)],  
-                                  [ self#statement (S.return_stmt c)]); comment}
+          { statement_desc = If (a, [self.statement self (S.return_stmt b)],  
+                                  [ self.statement self (S.return_stmt c)]); comment}
 
       | Return ({expression_desc = Seq _; _} as v) ->
           let block = Js_analyzer.rev_flatten_seq v  in
           begin match block with
           | {statement_desc = Exp last_one ; _} :: rest_rev
             ->  
-              super#statement 
-                (S.block (Ext_list.rev_map_append rest_rev [S.return_stmt last_one] (self#statement)))
+              super.statement self
+                (S.block (Ext_list.rev_map_append rest_rev [S.return_stmt last_one] (fun x -> self.statement self x)))
           | _ -> assert false
           end
       | Block [x]
           -> 
-            self#statement x 
-      | _ -> super#statement x 
-
-    method! block b =
+            self.statement self x 
+      | _ -> super.statement self x 
+    );      
+    block = fun self b ->
       match b with
       | {statement_desc = Block bs } :: rest ->
-          self#block ( bs @  rest)
+          self.block self ( bs @  rest)
       | x::rest  
         -> 
-          let st = self#statement x in 
-          let block = self#block rest in 
```

Comment:
in theory, we can tie the knot, fix `self` in the definition of `flatten_map` object